### PR TITLE
Add FFI function for batch signature removal.

### LINKED ISSUE #1006: Add feature to strip broken/unneeded key signatures
## Description

With the recent DoS attacks on keyservers, some applications have started to use stripped keys.
It would be good to have a way to give RNP a key with signatures, ask it to do a sanity check, and strip bad/broken or irrelevant signatures. Maybe I'd only want to keep the signatures made by one my own private keys.

### PR DESCRIPTION
**Pull request by ni4**
This PR adds FFI function `rnp_key_remove_signatures()`, which would allow to quickly filter out unneeded signatures on key export or import.
Fixes #1006 

Also it refactors some internal and tests code as a side effect.

### CODE REVIEW

File: include/rnp/rnp.h [hunk 1]
```diff
@@ -79,6 +79,17 @@ typedef uint32_t rnp_result_t;
 #define RNP_LOAD_SAVE_PERMISSIVE (1U << 8)
 #define RNP_LOAD_SAVE_SINGLE (1U << 9)
 
+/**
+ * Flags for the rnp_key_remove_signatures
+ */
+
+#define RNP_KEY_SIGNATURE_INVALID (1U << 0)
+#define RNP_KEY_SIGNATURE_UNKNOWN_KEY (1U << 1)
+#define RNP_KEY_SIGNATURE_NON_SELF (1U << 2)
```
Comment (dewyatt) [inline include/rnp/rnp.h:88]: I know these are *somewhat* self-explanatory, but I still think there's slight ambiguity so can we document these more?

Comment (ni4) [reply include/rnp/rnp.h:88]: @dewyatt Defines are documented in the description of `rnp_key_remove_signatures()`. Anyway, let's rename it to the `RNP_KEY_SIGNATURE_NON_SELF_SIG`, looks better.

File: include/rnp/rnp.h [hunk 2]
```diff
@@ -79,6 +79,17 @@ typedef uint32_t rnp_result_t;
...
+
+#define RNP_KEY_SIGNATURE_LEAVE (0U)
```
Comment (dewyatt) [inline include/rnp/rnp.h:90]: I would change this to `KEEP` rather than `LEAVE` if I understand what is doing correctly, it's a little clearer

Comment (ni4) [reply include/rnp/rnp.h:90]: Agree, sounds better.

**Comment by dewyatt** [comment]
Personally I'd like to see a more generic `key_walk_signatures` interface here.
Something like:

```cpp
typedef void (*rnp_key_walk_signatures_cb)(rnp_ffi_t              ffi,
                                      void *                 app_ctx,
                                      rnp_signature_handle_t sig);
rnp_result_t rnp_key_walk_signatures(rnp_key_handle_t      key,
                                     uint32_t              flags, // filters
                                     rnp_key_signatures_cb sigcb,
                                     void *                app_ctx);
```

Then the callback could inspect a signature, determine what it wants to do (remove it, print some info out, store some stuff, do nothing, ...) and make API calls to do this (`rnp_key_signature_remove` or something).

This would be more easily expanded in the future and more flexible.
I haven't reviewed the code, so I'm just commenting on this interface (maybe there's a reason to not do it this way).

Or an iterator-like interface would be nice to see too. Is there a reason to not do this?

**Comment by ni4** [comment]
> Personally I'd like to see a more generic `key_walk_signatures` interface here.
> Then the callback could inspect a signature, determine what it wants to do (remove it, print some info out, store some stuff, do nothing, ...) and make API calls to do this (`rnp_key_signature_remove` or something).
> 
> This would be more easily expanded in the future and more flexible.
> I haven't reviewed the code, so I'm just commenting on this interface (maybe there's a reason to not do it this way).
> 
> Or an iterator-like interface would be nice to see too. Is there a reason to not do this?

@dewyatt The main reason for the separate function is to be able to quickly cut-out all unneeded signatures. Removing those one-by-one would be of O(n^2) complexity, and say you have 1000 certifications on your main key and would like to export it with only the relevant self-sig. Or when you import large key and need to quickly clean it up.

However, we still may add separate `rnp_key_walk_signatures()` or iterator to be able to go through all the signatures and do whatever needed.

File: src/lib/pgp-key.h [hunk 3]
```diff
@@ -343,6 +348,34 @@ struct pgp_key_t {
...
+    bool is_revocation(const pgp_subsig_t &sig) const;
+
+    /** @brief Check whether signature is userid revocation */
+    bool is_uid_revocation(const pgp_subsig_t &sig) const;
+
+    /** @brief Check whether signature is subkey binding */
+    bool is_binding(const pgp_subsig_t &sig) const;
+
+    /**
+     * @brief Validate key-related signature, where this is a signing key.
```
Comment (antonsviridenko) [inline src/lib/pgp-key.h:373]: It is not clear what does this sentence mean.
I guess it is a replacement for `pgp_key_validate_signature()` which was documented this way:
> Validate key's signature.

Comment (ni4) [reply src/lib/pgp-key.h:373]: Thanks, will update the description.

**Comment by ni4** [comment]
@dewyatt @antonsviridenko Is it okay to go now?

**Review by antonsviridenko — APPROVED** [review]

**Review by dewyatt — APPROVED** [review]
The tests with stringstream seem a little weird to me but I assume there was good reason

**Comment by ni4** [comment]
> The tests with stringstream seem a little weird to me but I assume there was good reason

It was just the shortest way to write things down (and in one line). Not the best solution, but may be easily fixed afterwards anyway.

**Pull request merged**